### PR TITLE
Bad-BLE

### PR DESCRIPTION
**Pull request by yocvito**
Hi, I come here with a modification I made to flipper-Xtreme to add a basic bad BLE app. This is just the bad-USB app modified to work with BLE with a few modifications. 

# What has been done 

- Modified bad usb script thread to init/deinit bluetooth
- Modified bt, hal and gap layers to allow modifying Bt Adv name and Bt MAC address

# What remains to do

- [x] Integrate app to bad-USB
- [x] Fix applying keyboard layouts
- [x] Bt equivalent of `furi_hal_hid_get_led_state()` for `ducky_numlock_on()` to fix `ALTCHAR` and `ALTSTRING` commands
- [x] Remove pin verif procedure (done for IOS/OSX, seems not supported by Android/Linux/Windows: pin automatically accepted on flipper side)
- [ ] ~~Spoof Bt device type (if possible)~~ Very unlikely to be possible
- [x] ~~Look into forgetting paired bt keys~~ Saved to SD card instead
- [x] Maybe change app name to reflect having both USB and BT modes, `Bad-KB` perhaps?
- [x] Fix disconnect when switching script files

-----
# For the reviewer

- [ ] I've uploaded the firmware with this patch to a device and verified its functionality
- [ ] I've confirmed the bug to be fixed / feature to be stable

### CODE REVIEW

File: firmware/targets/f7/api_symbols.csv [hunk 1]
```diff
@@ -603,7 +608,7 @@ Function,+,byte_input_set_result_callback,void,"ByteInput*, ByteInputCallback, B
 Function,-,bzero,void,"void*, size_t"
 Function,-,calloc,void*,"size_t, size_t"
 Function,+,canvas_clear,void,Canvas*
-Function,+,canvas_commit,void,Canvas*
+Function,-,canvas_commit,void,Canvas*
```
Comment (ClaraCrazy) [inline firmware/targets/f7/api_symbols.csv:611]: why?

File: firmware/targets/f7/api_symbols.csv [hunk 2]
```diff
@@ -632,7 +637,7 @@ Function,+,canvas_glyph_width,uint8_t,"Canvas*, char"
 Function,+,canvas_height,uint8_t,Canvas*
 Function,-,canvas_init,Canvas*,
 Function,+,canvas_invert_color,void,Canvas*
-Function,+,canvas_reset,void,Canvas*
+Function,-,canvas_reset,void,Canvas*
```
Comment (ClaraCrazy) [inline firmware/targets/f7/api_symbols.csv:640]: same here, why?

File: firmware/targets/f7/api_symbols.csv [hunk 3]
```diff
@@ -995,13 +1000,17 @@ Function,+,furi_hal_bt_change_app,_Bool,"FuriHalBtProfile, GapEventCallback, voi
...
 Function,+,furi_hal_bt_get_key_storage_buff,void,"uint8_t**, uint16_t*"
+Function,?,furi_hal_bt_get_profile_adv_name,const char*,FuriHalBtProfile
+Function,?,furi_hal_bt_get_profile_mac_addr,const uint8_t*,FuriHalBtProfile
 Function,+,furi_hal_bt_get_radio_stack,FuriHalBtStack,
 Function,+,furi_hal_bt_get_rssi,float,
 Function,+,furi_hal_bt_get_transmitted_packets,uint32_t,
 Function,+,furi_hal_bt_hid_consumer_key_press,_Bool,uint16_t
 Function,+,furi_hal_bt_hid_consumer_key_release,_Bool,uint16_t
 Function,+,furi_hal_bt_hid_consumer_key_release_all,_Bool,
+Function,?,furi_hal_bt_hid_kb_free_slots,_Bool,uint8_t
```
Comment (ClaraCrazy) [inline firmware/targets/f7/api_symbols.csv:1013]: all of these `?` need to be `+` or `-`, depending on your needs.

File: firmware/targets/f7/api_symbols.csv (hunk 1 above)
Comment (ClaraCrazy) [inline firmware/targets/f7/api_symbols.csv:611]: ?

File: firmware/targets/f7/api_symbols.csv (hunk 2 above)
Comment (ClaraCrazy) [inline firmware/targets/f7/api_symbols.csv:640]: ?

**Review by ClaraCrazy — CHANGES_REQUESTED** [review]
See comments.

Also yes, this really should be integrated into BadUSB with a toggle. idk if @Willy-JL wants to help you with that, or maybe @floriangrousset but I doubt he has the time rn

**Comment by Willy-JL** [comment]
I have moved almost all the current code to the Bad USB app. There is now a toggle for which mode to use in the config pane of the app. From my testing it seems stable, but do let me know if you find issues with it since I had to use some workarounds to get the toggle to hide and show the BT-specific options on demand (namely having it be actually split in 2 different interfaces and switch between them when the option is toggled).

For now the toggle does not affect the actual functionality of Bad USB, it still works only over usb. The only code that remains to be ported over is actually the bt related code, in `bad_ble_script.c` to `bad_usb_script.c`. I have ported the new functions and definitions over and commented them out in `bad_ble_script.c`, along with the ones that were left unchanged. What remains to be done is to port the un-commented ones from `bad_ble_script.c` into Bad USB adding a check for whether to use USB or BT APIs to each of them. I did notice that these usually don't have access to the app/script context, so they wouldn't have direct access to the `is_bluetooth` variable I implemented on `BadUsbApp`; one possible solution that comes to mind is storing this value at script runtime in a module level variable, and having the APIs check that.

**Comment by Willy-JL** [comment]
Moved all the other code to the Bad USB app, should work properly now. One thing I noticed is `ducky_numlock_on()` still uses `furi_hal_hid_get_led_state()`, so that could cause issues... other than that I would say the first point in "What remains to do" is now done :smile:

**Comment by desktopsetup** [comment]
just cucked my wife feels good

**Comment by Willy-JL** [comment]
Fixed the keyboard layouts to be synced between usb and bt mode, and to reapply before starting a script since bt liked to have them reset now and then. Works perfectly so far for me. The issue with `furi_hal_hid_get_led_state()` and `ducky_numlock_on()` remains, so unless a bt equivalent of that is added the ALTSTRING and ALTCHAR commands will probably not work that well, if at all.

**Review by yocvito — COMMENTED** [review]
I added a check in gap layer to check for pairing method, and changes needed config to remove pin verif at pairing. I also modified bt init/deinit in bad usb script worker to modify the Bt HID profile pairing method at runtime.

**Pull request merged**